alexandria/skip_graph.py

```diff
@@ -0,0 +1,303 @@
+
+    nodes: Dict[Key, SGNode]
+
+    def __init__(self, initial_node: SGNode) -> None:
+        self.nodes = {initial_node.key: initial_node}
+
+    def insert(self, key: Key, current: SGNode) -> SGNode:
+        self.logger.debug("Inserting: %d", key)
+        left_neighbor, right_neighbor = self._search_insert_point(key, current, current.max_level)
+        self.logger.debug("Insertion point found: %s < %d < %s", left_neighbor, key, right_neighbor)
+        node = SGNode(key=key)
+        return self._insert_at_level(node, left_neighbor, right_neighbor, 0)
+
+    def _insert_at_level(self,
+                         node: SGNode,
+                         left: Optional[SGNode],
+                         right: Optional[SGNode],
+                         level: int) -> SGNode:
+        # Link the nodes on the current level
+        self._link_nodes(left, node, level)
+        self._link_nodes(node, right, level)
+
+        self.logger.debug("Level-%d: (%s < %s < %s)", level, left, node, right)
+
+        if left is None and right is not None:
+            assert right.get_left_neighbor(level) == node.key
+        if right is None and left is not None:
+            assert left.get_right_neighbor(level) == node.key
+        if left is not None and right is not None:
+            assert left.get_right_neighbor(level) == node.key
+            assert right.get_left_neighbor(level) == node.key
+
+        # Break if both neighbors are now null
+        if left is None and right is None:
+            self.nodes[node.key] = node
+            return node
+
+        # Now we iterate up the levels.  The `left` and
+        # `right` values are neighbors on the current level.  We check
+        # their membership vectors against the node's membership vector on the
+        # next level.  If they match, we link them to each other on the next
+        # level.  If not we scan through the neighbors on the current level
+        # until we either find one that matches on the next level **or** we run
+        # out of neighbors.
+        node_membership = node.get_membership_at_level(level + 1)
+
+        # Scan left until we find the first item in the list with a common membership vector
+        while left is not None and left.get_membership_at_level(level + 1) != node_membership:
+            # TODO: left neighbor needs to *maybe* have it's link on
+            # the next level set, but this requires scanning back in
+            # the other direction since it *might* link to something
+            # else on the next level.  This suggests we need to avoid
+            # populating the `null` levels and only have firm links.
+            next_left_key = left.get_left_neighbor(level)
+            if next_left_key is None:
+                left = None
+                break
+            else:
+                left = self.nodes[next_left_key]
+
+        while right is not None and right.get_membership_at_level(level + 1) != node_membership:
```

Comment:
Once we know our predecessor at `level + 1`, we don't have to scan through nodes until we find the correct right. The correct right will be `left.get_right_neighbor(level + 1)`!

Reply:
fancy.  nice catch/optimization.